File: unified-api/app/schemas/onboarding.py
"""Pydantic schemas for the onboarding flow."""

from __future__ import annotations

import uuid
from datetime import date

from pydantic import BaseModel


# ── Requests ────────────────────────────────────────────────────────


class ClassifyThemesRequest(BaseModel):
    etf_ids: list[uuid.UUID]


class CorrelationsRequest(BaseModel):
    etf_ids: list[uuid.UUID]
    lookback_days: int = 365


class CorrelatedPairInput(BaseModel):
    etf_id_a: uuid.UUID
    etf_id_b: uuid.UUID
    price_correlation: float | None = None
    holdings_overlap_pct: float | None = None


class AdvisorRequest(BaseModel):
    correlated_pairs: list[CorrelatedPairInput]
    all_etf_ids: list[uuid.UUID]


class PositionInput(BaseModel):
    etf_id: uuid.UUID
    shares: float
    entry_price: float
    entry_date: date
    invested_amount: float
    target_allocation: float | None = None


class ThemeInput(BaseModel):
    name: str
    color: str
    research_agent: str | None = None
    positions: list[PositionInput]


class OnboardingCompleteRequest(BaseModel):
    portfolio_name: str
    description: str | None = None
    themes: list[ThemeInput]


# ── Responses ───────────────────────────────────────────────────────


class ThemeClassification(BaseModel):
    label: str
    color: str
    etf_ids: list[uuid.UUID]
    etf_isins: list[str]
    research_agent: str | None = None


class ClassifyThemesResponse(BaseModel):
    themes: list[ThemeClassification]


class PairCorrelation(BaseModel):
    etf_id_a: uuid.UUID
    etf_id_b: uuid.UUID
    isin_a: str
    isin_b: str
    name_a: str
    name_b: str
    correlation: float


class PairOverlap(BaseModel):
    etf_id_a: uuid.UUID
    etf_id_b: uuid.UUID
    isin_a: str
    isin_b: str
    name_a: str
    name_b: str
    overlap_pct: float
    shared_holdings_count: int


class FlaggedPair(BaseModel):
    etf_id_a: uuid.UUID
    etf_id_b: uuid.UUID
    isin_a: str
    isin_b: str
    reason: str
    value: float


class CorrelationsResponse(BaseModel):
    price_correlations: list[PairCorrelation]
    holdings_overlaps: list[PairOverlap]
    flagged_pairs: list[FlaggedPair]


class RankedETF(BaseModel):
    etf_id: uuid.UUID
    isin: str
    name: str
    rank: int
    score_breakdown: dict


class PairRanking(BaseModel):
    pair_etf_ids: list[uuid.UUID]
    ranked_etfs: list[RankedETF]
    reasoning: str


class SuggestedETF(BaseModel):
    isin: str
    name: str
    ter: float | None = None
    vol_1y: float | None = None
    ret_risk_1y: float | None = None
    why: str


class ReplacementSuggestion(BaseModel):
    discard_etf_id: uuid.UUID
    theme: str
    suggested_etfs: list[SuggestedETF]
    reasoning: str


class AdvisorResponse(BaseModel):
    rankings: list[PairRanking]
    replacements: list[ReplacementSuggestion]


class OnboardingStatusResponse(BaseModel):
    is_onboarded: bool

File: frontend/src/types/onboarding.ts
// ── API Response Types (match backend schemas/onboarding.py) ────────

export interface ThemeClassification {
  label: string;
  color: string;
  etf_ids: string[];
  etf_isins: string[];
  research_agent: string | null;
}

export interface ClassifyThemesResponse {
  themes: ThemeClassification[];
}

export interface PairCorrelation {
  etf_id_a: string;
  etf_id_b: string;
  isin_a: string;
  isin_b: string;
  name_a: string;
  name_b: string;
  correlation: number;
}

export interface PairOverlap {
  etf_id_a: string;
  etf_id_b: string;
  isin_a: string;
  isin_b: string;
  name_a: string;
  name_b: string;
  overlap_pct: number;
  shared_holdings_count: number;
}

export interface FlaggedPair {
  etf_id_a: string;
  etf_id_b: string;
  isin_a: string;
  isin_b: string;
  reason: string;
  value: number;
}

export interface CorrelationsResponse {
  price_correlations: PairCorrelation[];
  holdings_overlaps: PairOverlap[];
  flagged_pairs: FlaggedPair[];
}

export interface RankedETF {
  etf_id: string;
  isin: string;
  name: string;
  rank: number;
  score_breakdown: Record<string, string>;
}

export interface PairRanking {
  pair_etf_ids: string[];
  ranked_etfs: RankedETF[];
  reasoning: string;
}

export interface SuggestedETF {
  isin: string;
  name: string;
  ter: number | null;
  vol_1y: number | null;
  ret_risk_1y: number | null;
  why: string;
}

export interface ReplacementSuggestion {
  discard_etf_id: string;
  theme: string;
  suggested_etfs: SuggestedETF[];
  reasoning: string;
}

export interface AdvisorResponse {
  rankings: PairRanking[];
  replacements: ReplacementSuggestion[];
}

export interface OnboardingStatusResponse {
  is_onboarded: boolean;
}

export interface HydrateETFsResponse {
  hydrated: number;
  already_populated: number;
  errors: string[];
}

// ── Local Wizard State Types ───────────────────────────────────────

export interface DraftETF {
  id: string;
  isin: string;
  name: string;
  ticker_yf: string | null;
}

export interface AllocationEntry extends DraftETF {
  mode: "weight" | "owned";
  target_weight: number;
  shares: number;
  entry_price: number;
  entry_date: string;
  theme_label: string;
  theme_color: string;
}

export type ResolutionChoice =
  | { type: "keep_winner"; winner_etf_id: string }
  | { type: "keep_and_replace"; winner_etf_id: string; replacement: SuggestedETF }
  | { type: "keep_both" };

export interface CorrelatedPairInput {
  etf_id_a: string;
  etf_id_b: string;
  price_correlation: number | null;
  holdings_overlap_pct: number | null;
}
